### PR TITLE
Prompt caching in `mlx_lm.server`

### PR DESCRIPTION
**Pull request by awni**
Added a basic prompt cache in `mlx_lm.server` for chat mode. But it does support chatting with cache reuse.

### CODE REVIEW

**Comment by chimezie** [comment]
This would not be back wards compatible with any later incorporation of batched input for generate (i.e., #948)

File: llms/mlx_lm/server.py [hunk 1]
```diff
@@ -474,15 +531,15 @@ def handle_completion(
 
     def handle_stream(
         self,
-        prompt: mx.array,
+        prompt: List[int],
```
Comment (chimezie) [inline llms/mlx_lm/server.py:534]: This in particular.  Can't we handle a single or multiple (batched) prompt that falls back to the behavior for a single prompt by default?

**Review by chimezie — CHANGES_REQUESTED** [review]
Leave door for  open for (#948)

File: llms/mlx_lm/server.py (hunk 1 above)
Comment (awni) [reply llms/mlx_lm/server.py:534]: We can always update this type to `List[List[int]]` when the time comes.

File: llms/mlx_lm/server.py [hunk 2]
```diff
@@ -496,16 +553,20 @@ def handle_stream(
         stop_sequence_suffix = None
         logging.debug(f"Starting stream:")
 
-        for (token, _), _ in zip(
+        prompt = self.get_prompt_cache(prompt)
+        prompt = mx.array(prompt)
```
Comment (angeloskath) [inline llms/mlx_lm/server.py:557]: I think this is a small efficiency bug for which you added the `mx.array` on line 562.

Basically, line 600 below casts this back to a list immediately so no need to cast to array here. We can cast only when passed to `generate_step` as you do below.

File: llms/mlx_lm/server.py [hunk 3]
```diff
@@ -559,7 +623,7 @@ def completion_usage_response(
     ):
         response = {
             "id": self.request_id,
-            "system_fingerprint": f"fp_{uuid.uuid4()}",
+            "system_fingerprint": self.system_fingerprint,
```
Comment (angeloskath) [inline llms/mlx_lm/server.py:626]: ❤️

File: llms/mlx_lm/server.py [hunk 4]
```diff
@@ -602,7 +665,7 @@ def handle_chat_completions(self) -> mx.array:
             prompt = convert_chat(body["messages"], body.get("role_mapping"))
             prompt = self.tokenizer.encode(prompt)
 
-        return mx.array(prompt)
+        return prompt
```
Comment (angeloskath) [inline llms/mlx_lm/server.py:668]: So this returns a list now? I am not sure why although there is no need to cast to mlx too early I guess. At the very least we probably need to change the return type annotation and docstring.

File: llms/mlx_lm/server.py [hunk 5]
```diff
@@ -614,11 +677,8 @@ def handle_text_completions(self) -> mx.array:
         # Determine response type
         self.request_id = f"cmpl-{uuid.uuid4()}"
         self.object_type = "text_completion"
-
         assert "prompt" in self.body, "Request did not contain a prompt"
-        prompt_text = self.body["prompt"]
-        prompt = self.tokenizer.encode(prompt_text)
-        return mx.array(prompt)
+        return self.tokenizer.encode(self.body["prompt"])
```
Comment (angeloskath) [inline llms/mlx_lm/server.py:681]: Same as above.

**Review by angeloskath — APPROVED** [review]
Looks fantastic! I left a few comments that may or may not need addressing.

**Pull request merged**